File: src/services/Services.DotNet/ToolExecutor.cs
namespace Services.DotNet;

public sealed class ToolExecutor : IToolExecutor
{
    private readonly IToolRegistry _toolRegistry;
    private readonly IReadOnlyDictionary<string, IToolHandler> _handlers;

    public ToolExecutor(IToolRegistry toolRegistry, IEnumerable<IToolHandler> handlers)
    {
        _toolRegistry = toolRegistry ?? throw new ArgumentNullException(nameof(toolRegistry));
        _handlers = (handlers ?? throw new ArgumentNullException(nameof(handlers)))
            .ToDictionary(handler => handler.ToolName, StringComparer.OrdinalIgnoreCase);
    }

    public IReadOnlyList<ToolDefinition> GetTools()
    {
        return _toolRegistry.GetTools();
    }

    public async Task<ToolExecutionResult> ExecuteAsync(
        string toolName,
        IReadOnlyDictionary<string, object?>? parameters = null,
        CancellationToken cancellationToken = default)
    {
        var tool = _toolRegistry.FindTool(toolName);

        if (!_handlers.TryGetValue(tool.Name, out var handler))
        {
            return new ToolExecutionResult
            {
                Success = false,
                Tool = tool.Name,
                Error = $"No application service handler registered for tool: {tool.Name}"
            };
        }

        var output = await handler.ExecuteAsync(parameters ?? new Dictionary<string, object?>(), cancellationToken);

        return new ToolExecutionResult
        {
            Success = true,
            Tool = tool.Name,
            Output = output
        };
    }
}

File: src/services/Services.NestJS/tool-executor.ts
import { ToolDefinition, ToolRegistry } from './tool-registry';

export type ToolParameters = Record<string, unknown>;

export interface ToolExecutionResult {
  success: boolean;
  tool: string;
  output?: unknown;
  error?: string;
}

export interface ToolHandler {
  toolName: string;
  execute(parameters: ToolParameters): Promise<unknown>;
}

export class ToolExecutor {
  private readonly handlers: Map<string, ToolHandler>;

  constructor(
    private readonly toolRegistry: ToolRegistry,
    handlers: ToolHandler[] = []
  ) {
    this.handlers = new Map(handlers.map((handler) => [handler.toolName, handler]));
  }

  getTools(): ToolDefinition[] {
    return this.toolRegistry.getTools();
  }

  async execute(toolName: string, parameters: ToolParameters = {}): Promise<ToolExecutionResult> {
    const tool = this.toolRegistry.findTool(toolName);
    const handler = this.handlers.get(tool.name);

    if (!handler) {
      return {
        success: false,
        tool: tool.name,
        error: `No application service handler registered for tool: ${tool.name}`,
      };
    }

    return {
      success: true,
      tool: tool.name,
      output: await handler.execute(parameters),
    };
  }
}